File: src/ds/E_Queue_01.java
package ds;

import ds.node.Node;

/**
 * start  11:01
 * end    10:08
 * @author Elena Novichkova
 *
 */

public class E_Queue_01 {
    private Node first;
    private Node last;

    public boolean isEmpty(){
        return first == null;
    }

    public int peek(){
        if(isEmpty())
            throw new RuntimeException("queue is empty");
        int data = first.data;
        return data;
    }

    public int remove(){
        int data = peek();
        if(last == first){
            first = null;
            last = null;
        } else {
            first = first.next;
        }
        return data;
    }

    public void add(int data){
        Node node = new Node(data);
        if(isEmpty()){
            first = node;
            last = node;
        } else {
            last.next = node;
            last = node;
        }
    }
}

File: src/ds/queue/E_Queue_01.js
import Node from "../node/Node.js";

const Queue = function(){
    this.top = null;
    this.bottom = null;

    this.isEmpty = function(){
        return this.top == null;
    }

    this.peek = function(){
        if(this.isEmpty()){
            throw new Error("Queue is empty");
        }
        return this.top.data;
    }

    this.remove = function(){
        let output = this.peek();
        if(this.top == this.bottom){
            this.top = null;
            this.bottom = null;
        } else {
            this.top = this.top.next;
        }
        return output;
    }

    this.add = function(value){
        let newNode =  new Node(value);
        if(this.isEmpty()){
            this.top = newNode;
            this.bottom = newNode;
        } else {
            this.bottom.next = newNode;
            this.bottom = newNode;
        }
    }
}

export default Queue;